### PR TITLE
DAOS-3846 Test: Automated daos test to verify device state for NVMe fault.

### PR DESCRIPTION
**Pull request by ravalsam**
Test cases will add small data set and generate NVMe fault on disk for rank1.
It will validate data set after object is getting rebuild.

During NVMe fault state will be changed from "FAULTY -> TEARDOWN" and
"TEARDOWN -> OUT", Verify this state in server log file.

Added Avocado test wrapper to run this test on weekly.

Test-tag-hw-medium: daos_test_nvme_recovery

Signed-off-by: Samir Raval <samir.raval@intel.com>

### CODE REVIEW

File: src/tests/suite/daos_test_common.c
Comment (NiuYawei) [inline src/tests/suite/daos_test_common.c:1132]: yeah, this could be a workaround before we have client API to query detailed information of individual pool target.

Comment (ravalsam) [reply src/tests/suite/daos_test_common.c:1132]: Correct, This is workaround until we get the API.

File: src/tests/ftest/daos_test/daos_core_test-nvme_recovery.yaml
Comment (ravalsam) [reply src/tests/ftest/daos_test/daos_core_test-nvme_recovery.yaml:58]: Yes, we run single test at a time  so server gets clean and restart and run from scratch. We can change and run all test in sequence once we have Hot-Plug available.

**Comment by ravalsam** [comment]
All Functional test failure are because of DAOS-5631 and not because of the changes made in this PR.

Newly added tests daos_test -N worked and passed.

**Review by NiuYawei — DISMISSED** [review]

File: src/tests/suite/daos_nvme_recovery.c
Comment (sydneyvanda) [inline src/tests/suite/daos_nvme_recovery.c:236]: First NORMAL->FAULTY transition should be observed. This is really the only useful transition to check for I would think in the logs. You can check for FAULTY->TEARDOWN as well, but the test might fail waiting for transition from TEARDOWN->OUT since we will not know how long to sleep for that. More useful to use a method to check the current blobstore state instead of looking into the logs (DAOS-4706) which I plan to work on next.

File: src/tests/suite/daos_nvme_recovery.c
Comment (sydneyvanda) [inline src/tests/suite/daos_nvme_recovery.c:267]: this can be done once $dmg storage replace command is added

File: src/tests/suite/daos_test_common.c
Comment (sydneyvanda) [reply src/tests/suite/daos_test_common.c:1132]: this is not for pool target, but for device state in log right?

**Comment by sydneyvanda** [comment]
I think this is ok for now, but will need to be updated once some of the work in DAOS-4705 is done, along with adding more daos recovery tests. Do we want to hold off and wait for some work in DAOS-4705 to get done before landing to avoid having to revise multiple tests?

File: src/tests/suite/daos_nvme_recovery.c
Comment (ravalsam) [reply src/tests/suite/daos_nvme_recovery.c:124]: Sure will do.

File: src/tests/suite/daos_nvme_recovery.c
Comment (ravalsam) [reply src/tests/suite/daos_nvme_recovery.c:236]: Let me add NORMAL->FAULTY check first.

For the rebuild time issue, I am writing small data set and waiting for 60 seconds to finish the rebuild which i think enough for now. We can hold If DAOS-4706 will be available soon otherwise let this get merged and we can change the method later. There are other tests which i am about to finish is depend on this one for dmg Api.

File: src/tests/suite/daos_nvme_recovery.c
Comment (ravalsam) [reply src/tests/suite/daos_nvme_recovery.c:267]: Right

File: src/tests/suite/daos_test_common.c
Comment (ravalsam) [reply src/tests/suite/daos_test_common.c:1132]: Correct, This is just to open server log file and check the specific state is exist.

**Comment by ravalsam** [comment]
@sydneyvanda 
There are two minor changes needed from your review comments. I have another PR ready in pipeline and I can do those two changes in new PR Instead of doing it here and waiting for days to get Ci result for this minor changes. Make sense ?

File: src/tests/suite/daos_nvme_recovery.c
Comment (sydneyvanda) [inline src/tests/suite/daos_nvme_recovery.c:199]: missing ioreq_fini() ?

Comment (ravalsam) [reply src/tests/suite/daos_nvme_recovery.c:199]: Yes let me add it, Thanks for catching it.

**Review by ravalsam — COMMENTED** [review]
Test-tag-hw-medium: daos_test_nvme_recovery

File: src/tests/suite/daos_test_common.c
Comment (sydneyvanda) [inline src/tests/suite/daos_test_common.c:1067]: this is not working, instead I would suggest just coping everything after -o or --config returned after ps aux command.

File: src/tests/suite/daos_test_common.c
Comment (sydneyvanda) [inline src/tests/suite/daos_test_common.c:1124]: this will cause a problem if the pid is appended to the logfile automatically (ie the D_LOG_FILE_APPEND_PID=1). so this option will need to be checked, and then the logfile name appended to include the PID returned from the ps aux command in get_server_config()

File: src/tests/suite/daos_nvme_recovery.c
Comment (sydneyvanda) [inline src/tests/suite/daos_nvme_recovery.c:279]: I don't really think lookup and verify of all records is required for this specific test. I suggest simplifying the test to just verifying the states and no the data.
Also, when running this with 1 server and 2 disks, the verification fails so something is wrong there anyways
Found state NORMAL -> FAULTY in Log file  /tmp/smvanda/daos_server.log     # map to D_LOG_FILE=/tmp/server.log
Found state FAULTY -> TEARDOWN in Log file  /tmp/smvanda/daos_server.log     # map to D_LOG_FILE=/tmp/server.log
Found state TEARDOWN -> OUT in Log file  /tmp/smvanda/daos_server.log     # map to D_LOG_FILE=/tmp/server.log
Insert fresh record with 100 extents
Lookup and Verify records:
[  ERROR   ] --- 0xfffffffffffffc13 != 0
[   LINE   ] --- src/tests/suite/daos_obj.c:453: error: Failure!

**Review by sydneyvanda — CHANGES_REQUESTED** [review]

File: src/tests/suite/daos_test_common.c
Comment (ravalsam) [reply src/tests/suite/daos_test_common.c:1067]: Updating the code to pick the next string after -o and --config

File: src/tests/suite/daos_test_common.c
Comment (ravalsam) [reply src/tests/suite/daos_test_common.c:1124]: This is tricky as we also need to match daos_io_server -I 0 and -I 1 with ranks as some time rank1 process starts first.

We will be skipping this for now and will land another PR where we do not have to read the stats from File instead from dmg command (PR will be created by @sydneyvanda)

File: src/tests/suite/daos_nvme_recovery.c
Comment (ravalsam) [reply src/tests/suite/daos_nvme_recovery.c:279]: Will remove the data verification.

File: src/tests/suite/daos_nvme_recovery.c
Comment (sydneyvanda) [inline src/tests/suite/daos_nvme_recovery.c]: This can be left in the test, so faulty reaction will have something to rebuild when the device is marked as faulty. I was just suggesting to remove the additional insert fresh record section when rank < 1, and also the record verification at the end

**Review by sydneyvanda — DISMISSED** [review]
just one additional comment, but since the daos_nvme_recovery test will be updated after the mgmt blobstore state API (DAOS-4706) lands, no need to update in this patch

**Review by rpadma2 — APPROVED** [review]

File: src/common/tests_dmg_helpers.c
Comment (jolivier23) [inline src/common/tests_dmg_helpers.c:632]: D_ALLOC_PTR ?

File: src/common/tests_dmg_helpers.c
Comment (jolivier23) [inline src/common/tests_dmg_helpers.c:691]: Mixing calloc with D_FREE will break the cause a false leak in log parsing tool

File: src/tests/suite/daos_nvme_recovery.c
Comment (jolivier23) [inline src/tests/suite/daos_nvme_recovery.c:228]: we do this in a lot of places but this would leak memory if it triggered.   I don't think you need to fix it here, more of a comment.

File: src/tests/suite/daos_test_common.c
Comment (jolivier23) [inline src/tests/suite/daos_test_common.c:1077]: It would be good to print the stderror(errno) here or something so we have a clue why it failed.

File: src/tests/suite/daos_nvme_recovery.c
Comment (ravalsam) [reply src/tests/suite/daos_nvme_recovery.c]: ok I will update the code in next PR.

File: src/tests/suite/daos_test_common.c
Comment (jolivier23) [inline src/tests/suite/daos_test_common.c:1110]: probably should just use free here since getline is the allocator.  Perhaps we should add a D_GETLINE at some point to appropriately log it.  But for now, we shouldn't pair D_FREE with glibc allocation

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/tests/suite/daos_test_common.c
Comment (ravalsam) [reply src/tests/suite/daos_test_common.c:1077]: Popen internally performed by shell so any failure error will be printed on console.
It will return NULL in case of fork(), pipe() or memory allocation failed which is just to check it's not NULL in any case before reading it through in case of command success.

File: src/common/tests_dmg_helpers.c
Comment (ashleypittman) [reply src/common/tests_dmg_helpers.c:691]: Jeff is correct, and we should use the macros but I don't think the unit tests are using the log parsing tool.

**Review by sydneyvanda — APPROVED** [review]

**Comment by ravalsam** [comment]
@daos-stack/daos-gatekeeper 
Test cases which was added by this PR got passed.
There are failure but those are related to existing defect DAOS-5755 and DAOS-5591.
Some of the test failed on CetOD7/LEAP 15 are not related to this change.

**Comment by ravalsam** [comment]
We have few errors which are related to existing defects DAOS-5745, DAOS-5767. DaosCoreTestRebuild.01 test got timeout but it was working on previous run before merge the master. so looks like issue in master but not from changes made from this PR.

**Review by mjmac — APPROVED** [review]
It looks fine to me... Seems like a lot of C to maintain for a test that could have more easily been written in python, though.

**Comment by ravalsam** [comment]
> It looks fine to me... Seems like a lot of C to maintain for a test that could have more easily been written in python, though.

Thanks for the review. The purpose is to have with C so it can be used in daos_test. There are some more test cases will be developed based on this and expand daos_test for NVMe disk fault.

**Pull request merged**
